lib/cfg.js

```diff
@@ -0,0 +1,368 @@
+// Copyright (c) 2012-2017, Najjar Chedy
+//
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+//     * Redistributions of source code must retain the above copyright notice,
+//       this list of conditions and the following disclaimer.
+//     * Redistributions in binary form must reproduce the above copyright
+//       notice, this list of conditions and the following disclaimer in the
+//       documentation and/or other materials provided with the distribution.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+
+var _ = require('underscore-node');
+
+function ControlFlowGraph(rules){
+    this.rules = rules;
+    
+}
+
+function seperateCodeFromData(asmArr) {
+    
+    var isCode = function(asmArr, index){
+        return (((asmArr[index].source !== null) &&
+                (asmArr[index+1].source !== null)) ||
+                ((asmArr[index].source === null) &&
+                 (asmArr[index+1].source !== null)) ||
+                 ((asmArr[index].source !== null) &&
+                 (asmArr[index+1].source === null)));
+    };
+
+    var code = [];
+    var first = 0;
+    var last = asmArr.length;
+    
+    if(first == last) return code;
+
+    
+    while(first != last-1){
+        if(isCode(asmArr, first))
+            code.push(JSON.parse(JSON.stringify(asmArr[first])));
+        
+        ++first;
+    }
+    if(asmArr[first].source){
+        code.push(JSON.parse(JSON.stringify(asmArr[first])));
+    }
+
+    return code;
+}
+
+    var gccX86 = {
+                filterData:  seperateCodeFromData,
+                isFunctionEnd:function(x)  { return ((x[0] != ' ') && (x[0] != '.') &&
+                                                             (x.indexOf(':') != -1)) ;},
+
+                isBasicBlockEnd: function(x) { return x[0] == ".";} ,
+
+                getInstructionType:function(inst) {
```

Comment:
also worth noting this only supports x86. We are trying to add support for other processors into other areas e.g. the tooltips. This is fine for now but we should probably try and come up with a nice unified solution to this stuff.

Reply:
And this might be a good fit for regular expressions, to make it easier to add other architectures. Maybe...you know how regular expressions can be trouble...

Reply:
I will discuss this with you later